src/HttpClient/HttpClientFactory.php

```diff
@@ -0,0 +1,116 @@
+use Http\Client\Common\Plugin\HeaderSetPlugin;
+use Http\Client\Common\Plugin\RetryPlugin;
+use Http\Client\Common\PluginClient;
+use Http\Client\Curl\Client as HttpCurlClient;
+use Http\Client\HttpAsyncClient as HttpAsyncClientInterface;
+use Http\Discovery\HttpAsyncClientDiscovery;
+use Http\Message\StreamFactory;
+use Http\Message\UriFactory as UriFactoryInterface;
+use Sentry\HttpClient\Authentication\SentryAuthentication;
+use Sentry\HttpClient\Plugin\GzipEncoderPlugin;
+use Sentry\Options;
+
+/**
+ * Default implementation of the {@HttpClientFactoryInterface} interface that uses
+ * Httplug to autodiscover the HTTP client if none is passed by the user.
+ */
+final class HttpClientFactory implements HttpClientFactoryInterface
+{
+    /**
+     * @var UriFactoryInterface The PSR-7 URI factory
+     */
+    private $uriFactory;
+
+    /**
+     * @var StreamFactory The PSR-7 stream factory
+     */
+    private $streamFactory;
+
+    /**
+     * @var HttpAsyncClientInterface|null The HTTP client
+     */
+    private $httpClient;
+
+    /**
+     * @var string The name of the SDK
+     */
+    private $sdkIdentifier;
+
+    /**
+     * @var string The version of the SDK
+     */
+    private $sdkVersion;
+
+    /**
+     * Constructor.
+     *
+     * @param UriFactoryInterface           $uriFactory    The PSR-7 URI factory
+     * @param StreamFactory                 $streamFactory The PSR-17 stream factory
+     * @param HttpAsyncClientInterface|null $httpClient    The HTTP client
+     * @param string                        $sdkIdentifier The SDK identifier
+     * @param string                        $sdkVersion    The SDK version
+     */
+    public function __construct(UriFactoryInterface $uriFactory, StreamFactory $streamFactory, ?HttpAsyncClientInterface $httpClient, string $sdkIdentifier, string $sdkVersion)
+    {
+        $this->uriFactory = $uriFactory;
+        $this->httpClient = $httpClient;
+        $this->sdkIdentifier = $sdkIdentifier;
+        $this->sdkVersion = $sdkVersion;
+        $this->streamFactory = $streamFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create(Options $options): HttpAsyncClientInterface
+    {
+        if (null === $options->getDsn()) {
+            throw new \RuntimeException('Cannot create an HTTP client without the Sentry DSN set in the options.');
+        }
+
+        $httpClient = $this->httpClient;
+
+        if (null !== $httpClient && null !== $options->getHttpProxy()) {
+            throw new \RuntimeException('The "http_proxy" option does not work together with a custom HTTP client.');
+        }
+
+        if (null === $httpClient && null !== $options->getHttpProxy()) {
+            if (!class_exists(HttpCurlClient::class)) {
+                throw new \RuntimeException('The "http_proxy" option requires the `php-http/curl-client` package to be installed.');
+            }
+
+            $httpClient = new HttpCurlClient(null, null, [
```

Comment:
This piece of code will conflict with #890.

Reply:
Indeed, depending on which one will be merged first I will rebase the second PR